tests/aws/services/s3/test_s3_api.py

```diff
@@ -1869,3 +1873,252 @@ def test_put_object_if_none_match_versioned_bucket(self, s3_bucket, aws_client,
+            aws_client.s3.complete_multipart_upload(
+                Bucket=s3_bucket,
+                Key=key,
+                MultipartUpload={"Parts": parts},
+                UploadId=upload_id,
+                IfMatch=put_obj_etag_1,
+            )
+        snapshot.match("complete-multipart-if-match-put-before", e.value.response)
+
+        # the previous PutObject request was done between the CreateMultipartUpload and completion, so it takes
+        # precedence
+        # you need to restart the whole multipart for it to work
+        with pytest.raises(ClientError) as e:
+            aws_client.s3.complete_multipart_upload(
+                Bucket=s3_bucket,
+                Key=key,
+                MultipartUpload={"Parts": parts},
+                UploadId=upload_id,
+                IfMatch=put_obj_etag_2,
+            )
+        snapshot.match("complete-multipart-if-match-put-during", e.value.response)
+
+        create_multipart = aws_client.s3.create_multipart_upload(Bucket=s3_bucket, Key=key)
+        snapshot.match("create-multipart-again", create_multipart)
+        upload_id = create_multipart["UploadId"]
+
+        upload_part = aws_client.s3.upload_part(
+            Bucket=s3_bucket, Key=key, UploadId=upload_id, Body="test", PartNumber=1
+        )
+        parts = [{"ETag": upload_part["ETag"], "PartNumber": 1}]
+
+        complete_multipart = aws_client.s3.complete_multipart_upload(
+            Bucket=s3_bucket,
+            Key=key,
+            MultipartUpload={"Parts": parts},
+            UploadId=upload_id,
+            IfMatch=put_obj_etag_2,
+        )
+        snapshot.match("complete-multipart-if-match-put-before-restart", complete_multipart)
+
+    @markers.aws.validated
+    def test_multipart_if_match_with_put_identical(self, s3_bucket, aws_client, snapshot):
+        key = "test-precondition"
+        put_obj = aws_client.s3.put_object(Bucket=s3_bucket, Key=key, Body="test")
+        snapshot.match("put-obj", put_obj)
+        put_obj_etag_1 = put_obj["ETag"]
+
+        create_multipart = aws_client.s3.create_multipart_upload(Bucket=s3_bucket, Key=key)
+        snapshot.match("create-multipart", create_multipart)
+        upload_id = create_multipart["UploadId"]
+
+        upload_part = aws_client.s3.upload_part(
+            Bucket=s3_bucket, Key=key, UploadId=upload_id, Body="test", PartNumber=1
+        )
+        parts = [{"ETag": upload_part["ETag"], "PartNumber": 1}]
+
+        put_obj_2 = aws_client.s3.put_object(Bucket=s3_bucket, Key=key, Body="test")
+        snapshot.match("put-obj-during", put_obj_2)
+        # same ETag as first put
+        put_obj_etag_2 = put_obj_2["ETag"]
+        assert put_obj_etag_1 == put_obj_etag_2
+
+        # it seems that even if we overwrite the object with the same content, S3 will still reject the request if a
+        # write operation was done between creation and completion of the multipart upload, like the `Delete`
+        # counterpart of `IfNoneMatch`
```

Comment:
question: because every write changes the ETag I assume?

Reply:
almost, if the content of the object is the same, the ETag does not change, it's most of the time a MD5 hash of the object content. Which is why it's a bit surprising! I also gave a quick write up here: https://github.com/apache/arrow-rs/issues/6799#issuecomment-2502344533

This behavior was documented for `IfNoneMatch`, but not for `IfMatch` it seems, but I suspect it's because they use the same logic/primitive for both.